parcels/kernel.py

```diff
@@ -379,22 +379,22 @@ def remove_deleted(pset):
         remove_deleted(pset)
 
         # Identify particles that threw errors
-        error_particles = [p for p in pset.particles if p.state not in [ErrorCode.Success, ErrorCode.Evaluate]]
-
-        while len(error_particles) > 0:
+        error_particles = np.isin(pset.particle_data['state'], [ErrorCode.Success, ErrorCode.Evaluate], invert=True)
+        while np.any(error_particles):
             # Apply recovery kernel
-            for p in error_particles:
-                if p.state == ErrorCode.Repeat:
-                    p.reset_state()
-                elif p.state in recovery_map:
-                    recovery_kernel = recovery_map[p.state]
-                    p.state = ErrorCode.Success
-                    recovery_kernel(p, self.fieldset, p.time)
-                    if(p.isComputed()):
-                        p.reset_state()
+            for p in np.where(error_particles)[0]:
+                particles.set_index(p)
+                if particles.state == ErrorCode.Repeat:
+                    particles.state = ErrorCode.Evaluate
+                elif particles.state in recovery_map:
+                    recovery_kernel = recovery_map[particles.state]
+                    particles.state = ErrorCode.Success
+                    recovery_kernel(particles, self.fieldset, particles.time)
+                    if particles.state == ErrorCode.Success:
+                        particles.state = ErrorCode.Evaluate
                 else:
                     logger.warning_once('Deleting particle because of bug in #749 and #737')
-                    p.delete()
+                    particles.delete()
```

Comment:
ah, now I see why in sao my nice state-access functions don't work - they need to be now member functions of the ParticleSet, not the Particle itself. With **delete**, that switch from `Particle` member function to `ParticleSet` member function worked cause there where mutual callers in those classes. We could pretty simply switch back to the member functions rather than directly setting the state, by just introducing the functions `isComputed()`, `reset_state()` and `succeeded()` to the ParticleSet. But it's just a suggestion - if you think it's fine to externally allow access to the state variable, then that's ok.

Reply:
I understand yes. Well, perhaps you can create a separate PR to implement this?

Reply:
I've now implemented a general `Particle.set_state(res)` function